### ISSUE TITLE
unable to init a lambda using a profile that contains a role_arn

### ISSUE THREAD

**Issue by luck02** [apex/apex#524]
- __Operating system__:  OSX 10.11.6
  - __Apex Version__: 0.10.2

As far as I can tell my credentials and config file are correct.  when I execute:

```shell
⇒  aws iam create-role --profile data --role-name test-this-role-delete-when-done --assume-role-policy-document file://policy-doc.json
```

I get the expected output (role is created).  note I'm using the --profile data flag.

So I'm going to assume my AWS config is demonstrably correct

When I execute:

```shell
⇒  apex --profile data init -ldebug


             _    ____  _______  __
            / \  |  _ \| ____\ \/ /
           / _ \ | |_) |  _|  \  /
          / ___ \|  __/| |___ /  \
         /_/   \_\_|   |_____/_/\_\



  Enter the name of your project. It should be machine-friendly, as this
  is used to prefix your functions in Lambda.

    Project name: proj_name

  Enter an optional description of your project.

    Project description:

  [+] creating IAM ub-data-event-bus-persistor_lambda_function role
   ⨯ Error: creating role: NoCredentialProviders: no valid providers in chain. Deprecated.
       	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

When I execute this command without the --profile data flag I get:

```shell
  [+] creating IAM ub-data-event-bus-persistor_lambda_function role
   ⨯ Error: creating role: AccessDenied: User: arn:aws:iam::99999999:user/gary.lucas is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::99999999:role/lambda_user
       	status code: 403, request id: 93617ee6-63a6-11e6-a945-c1fc127febbb
```

My default role does not have those required permissions so that's expected.

My credentials file is standard and set up via aws configure command.  My config file looks more or less like this:
```shell
[default]
region = us-east-1
output = json

[profile data]
role_arn = arn:aws:iam::999999:role/a-valid-cross-accountRole
source_profile = default
output = json
region = us-east-1
```

Note that I had to add region to profile data as apex wasn't picking it up.

My assumption (untested) is that apex is just issueing commands via the AWS SDK so it should work.  But perhaps that's not the case?  Does anyone have an experience using roles that worked?

**Comment by DavidTPate** [comment]
Hi there! You're correct in your assumption that Apex is just invoking that AWS SDK for Go to setup roles, retrieve them, etc. I think what is going on is that the AWS SDK for Go doesn't support assuming a role unless the role comes from EC2 metadata.

Could you try doing the same thing but instead of using a role use hardcoded credentials, ie:

```
[profile data]
aws_access_key_id = accessKey
aws_secret_access_key = secret
region = us-east-1
```

I'm guessing this is going to work, though I understand that is not what you desire to do so that you don't have to have hardcoded keys. But if this works then we'll at least know where the issue lies and we can see what can be done next.

**Comment by luck02** [comment]
Ok, interesting.  I'm just trying to figure out what the minimum required set of permissions required for Apex to manage a lambda.  It's fairly extensive.  Role creation + lambda management.  

Thanks for the quick reply, that's super helpfull and very much appreciated!

**Comment by DavidTPate** [comment]
@luck02 Yeah, we don't seem to have the Minimum/Suggested Role Policy for the Role/Credentials used by Apex anywhere. I'll see about compiling those.

**Comment by DavidTPate** [comment]
@luck02 In #529 I put together a set of what I'm pretty sure is the absolute minimum policy for use with running Apex to manage Lambda. I tested all the functionality that I'm aware of within Apex with it and after adding a few that I missed it works as expected.

Please let me know if you hit any permissions that I missed 👍

**Comment by luck02** [comment]
Thanks a ton :)

**Comment by mikereinhold** [comment]
I see the same issue with my cross account roles. 

From what I can tell, the aws-sdk-go properly supports assuming roles as of [v1.3.0](https://github.com/aws/aws-sdk-go/issues/472#issuecomment-237342876)

Terraform also has implemented assume role functionality, see [Terraform PR 8638](https://github.com/hashicorp/terraform/pull/8638). Terraform's scenario is more complex as you can specify the role and source profile in the Terraform provider configuration.

From what I can tell, you just have to use the stsCreds.AssumeRoleProvider as part of the provider chain. For Apex's use case, you should simply define the credentials chain explicity and include the AssumeRoleProvider instead of relying only on the default SharedCredentialsProvider (root.go:113?)...

**Comment by hybby** [comment]
Seeing the same issue here when using a profile with `role_arn` instead of straight-up credentials.  Apex version 0.10.3.

**Comment by maxmanders** [comment]
Nice solution @hybby - I'm also keen to see this facility adopted in Apex.

**Comment by jhg-tgu** [comment]
Hi, I have issue using a role-arn profile:
```   ⨯ Error: function ******: NoCredentialProviders: no valid providers in chain. Deprecated. 
        For verbose messaging see aws.Config.CredentialsChainVerboseErrors```
I'm using Apex 0.12.0 and, I don't know in Go, but in Python it's possible with AWS SDK use assume role and connect with role-arn, I thing it's a useful feature for Apex.

**Comment by jhg-tgu** [comment]
@hybby using your solution I have another error:
```
   ⨯ Error: function ******: AccessDeniedException: Cross-account pass role is not allowed.
        status code: 403, request id: bd8cb220-e716-11e6-839c-f3e4dce9dc6d
```

**Comment by ryantuck** [comment]
```
$ export AWS_SDK_LOAD_CONFIG=1
$ apex --profile assumed_profile init
...
```

worked for me.

**Comment by victorcoder** [comment]
@ryantuck solution doesn't work for me. 😞 

Only way to make it work is via wrapper for AWS creds as @hybby said
